### ISSUE TITLE
6주차 Gaussian Processes and Bayesian Optimization

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/bayesian-methods-in-machine-learning#4]
참조 

* [nonparametric-methods.pdf](https://github.com/sweaterr/bayesian-methods-in-machine-learning/files/1730282/nonparametric-methods.pdf)
* [gaussian-processes.pdf](https://github.com/sweaterr/bayesian-methods-in-machine-learning/files/1730281/gaussian-processes.pdf)
* [gp-for-machine-learning.pdf](https://github.com/sweaterr/bayesian-methods-in-machine-learning/files/1730286/gp-for-machine-learning.pdf)
* [Nuance of GP.pdf](https://github.com/sweaterr/bayesian-methods-in-machine-learning/files/1730283/Nuance.of.GP.pdf)

* http://www.robots.ox.ac.uk/~mebden/reports/GPtutorial.pdf

**Comment by sweaterr** [comment]
### Nonparametric methods
* [nonparametric-methods.pdf](https://github.com/sweaterr/bayesian-methods-in-machine-learning/files/1718374/nonparametric-methods.pdf)
* 파라메터가 데이터 사이즈에 depend 한다.
  * K-nearest neighbors
  * Nadaraya-watson 좀더 smooth

* Kernels
  * Gaussian
  * Uniform

*  Parametric vs Non-paprametric
    * parametric
      * 모델 복잡도에 한계가 있다.
      * 추론이 빠르다.
      * 학습이 느리다.
    *  Non-parametric
       * 모델 복잡도에 한계가 없다.
       * 추론이 느리다.
       * 학습은 모든 데이터를 기억하는 것이다.

**Comment by sweaterr** [comment]
### Gaussian Process
* 랜덤 프로세스
  * 한 차원 더 있음
  *  $\mathbb{R}^d$ 인 벡터를 뽑고 랜덤 변수 $f(x)$를 할당한다.
  * d=1이면 시간같은것 일수 있다.
* $\mathbb{R}^d$ 인 벡터를 뽑고 각 원소에 대해서 가우시안 f를 씌운 joint distribution도 역시 가우시안이다. 
* stationary process
  * 유한-차원 분포가 점의 상대적 위치에만 depend 한다.
  * 주기성이 있는 것은 stationary 하지 않다.
    * joint distribution은 공간의 다른 부분에서 다르다.
  * trend가 있는 것은  stationary 하지 않다.
    * 샘플을 뽑아 평균내면 그 위치가 어딘지 알 수 있다.
* stationary process 는 trend가 없어야 한다.
* Gaussian process is stationary 
  * $m(x) = const $
  * $K(x_1,x_2) = \tilde{K}(x_1 - x_2) $ 
  * Var[f(x)] = $\tilde{K}(0)$
  *  ![image](https://user-images.githubusercontent.com/1518919/36246011-18d8b51e-1271-11e8-8b71-234f18ab0157.png)
* 
![image](https://user-images.githubusercontent.com/1518919/36246024-268903c6-1271-11e8-8ad7-58b632c4cf7f.png)

**Comment by sweaterr** [comment]
### GP for maching learning
* 보통 리그레션과 달리 full posterior $p(f(x)|f(x_1),...,f(x_n))$을 모델링한다.

<img src="https://user-images.githubusercontent.com/1518919/36272942-4dcfecec-12c6-11e8-949a-c09e7ecff9dd.png" width="400" height="200" />
<img src="https://user-images.githubusercontent.com/1518919/36273005-7c44d0a6-12c6-11e8-9180-06ae2be774c1.png" width="400" height="100" />

* $=\mathcal{N}(f(x)|\mu,\sigma^2)$
  * $\mu = k^TC^{-1}f$
  * $\sigma^2 = K(0) - k^TC^{-1}k$, $k^TC^{-1}k$는 데이터를 보고 나서 얼마나 variance가 줄어드는가를 나타냄 

<img src="https://user-images.githubusercontent.com/1518919/36273604-5ad26350-12c8-11e8-9715-17c42ca07176.png" width="300" height="200" />

* 데이터가 있는 지점의 variance는 0이다.
* 데이터에서 벗어날수록 variance가 증가한다.
* 데이터에서 많이 벗어나면 Process가  stationary하게 된다. 그래서 평균은 0이 되고 분산은 초기분산과 같아진다($K(0)$).

전처리
* Far from data: $\mathop{\mathbb{E}}f(x) = 0$, $Var[f(x)] = K(0)$
* 트렌드와 주기를 없앤다.
* 평균을 빼고 노말라이즈 한다.
* 새로운 data point를 예측할 때는 전처리를 반대로 대돌려야 한다.

**Comment by sweaterr** [comment]
### Nuances of GP (1/2)
* [Nuance of GP.pdf](https://github.com/sweaterr/bayesian-methods-in-machine-learning/files/1730247/Nuance.of.GP.pdf)
* GP를 훈련하는 유용한 트릭을 소개한다.

<img src="https://user-images.githubusercontent.com/1518919/36274585-882d8188-12cb-11e8-8ea8-4c0dd75db8b4.png" width="300" height="200" />

*  노이지한 데이터에 GP를 맞춘 데이터
* $\hat{f}(x) = f(x) + \epsilon$
* $\epsilon \sim \mathcal{N}(0,s^2)$
* $m(x)=0$: 이 경우 랜덤 프로세스의 평균은 0이다.
  * 평균이 0인 $f(x)$와 $\epsilon$을 더했기 때문에
* $\hat{K}(x_i-x_j)=K(x_i-x_j)+s^2\mathbb{I}[x_i=x_j]$
  * $s^2\mathbb{I}[x_i=x_j]$는 다른 포지션에 있는 노이지 샘플은 covariance가 없다는 뜻

<img src="https://user-images.githubusercontent.com/1518919/36293600-516f7432-131e-11e8-985c-5449af8755ba.png" width="300" height="200" />

* variance 0인 것이 사라짐
* mean도 좀 더 smooth 해짐

Kernel paramters
* Kernel paramters를 조절하면 좀 더 잘 할 수 있다. 
* RBF에서는 $l$을 조절
* gradient descent로 자동으로 조절할 수도 있다.
  * $p(f(x_1),f(x_2),...,f(x_n)|\sigma^2,l,s^2) \rightarrow \text{max}_{\sigma^2,l,s^2}$ 
  * $\mathcal{N}
(f(x_1),f(x_2),...,f(x_n)|0,C) \rightarrow \text{max}_{\sigma^2,l,s^2}$ 
*  노이지한 데이터에 맞추면 $s_2$가 커짐

**Comment by sweaterr** [comment]
### Nuances of GP (2/2)
Classifcation
* 자세히 다루지는 않는다 일단 스킵

Inducing inputs
* GP는 계산 복잡도가 크다. 제일큰건 covariance matrix를 역행렬 취하는데 드는 비용이 $\mathcal{O}(n^3)$
* svm의 support vector 처럼 전체 데이터셋이 아니라 m개의 작은 데이터셋만 사용하자.
* 품질이 그다지 떨어지지 않음

**Comment by sweaterr** [comment]
### Bayesian optimization
* GP를 optimization에도 쓸 수 있다.

Black-box optimization
* $f(x) \rightarrow max_x$ 
* gradient를 모르면 gradient를 근사하거나, grid search/random search 를 한다.
* 그러나 샘플하나를 뽑는 것이 매우 비싼 경우가 있다.
  * 어디에 기름이 있는지 알아내는 문제, 1 sample=$1,000,000
  * 뉴럴넷의 hyperparameter를 알아내는 문제. 1sample=10hours
* $f(x) \rightarrow max_x$ 를 최소 시도로 알아내자
* surrogate model: $\hat{f} \approx f$
  * 평가하는데 cheap 해야한다.
  * 복잡한 함수를 모델링할 수 있어야 하고(Nonparametic이면 좋음), 불확실성을 추정할 수 있어야 하므로 GP가 좋은 선택이다.
* Acquisition function: $\mu(x)$
  * 최적화 이익을 추정한다.
  * surrogate model을 사용한다.
  * Exploration과 Exploitation 사이의 균형을 맞춰야 한다.
  * Maximum probability of improvement(MPI)
    * $f^*$는 current best value
    * $\mu(x) = P(\hat{f}(x) \geq f^* + \epsilon) = \Phi(\frac{\mathbb{E}\hat{f}(x)-f^*-\epsilon}{Var[\hat{f}(x)]})$ 
      * 맨 오른쪽은 정규분포 
  * Upper confidence bound(UCB)
    * $\mu(x) = \mathbb{E}\hat{f}(x) + \eta Var[\hat{f}(x)]$ 
  *  Expected improvement(EI)
    * $\mu(x) = \mathbb{E} \max(f(x)-f^*,0)=Var[\hat{f}(x)] \cdot [z\Phi(z)+\phi(z)]$  
    * $z=\frac{\mathbb{E}\hat{f}(f)(x)-m(x)}{Var[\hat{f}(x )]}$

Example
* 몇 개의 점에서 시작
* 수렴할 때까지 다음을 반복
  * GP를 훈련
  * gd를 사용해서 $\mu(x)$의 최대점을 찾음
  * $\mu(x)$의 최대점에서 함수를 평가 

Random search vs Gaussian processes
* RS는 병렬화가 쉬지만, 대신 많이 시도해야 함 
* GP는 계산시간이 걸리기 때문에, 한번 평가가 비싼 작업에 적합하다.

**Comment by sweaterr** [comment]
### Application of Bayesian optimization
* 뉴럴넷 하이퍼파라메터 튜닝 
  * 레이어수, BN, learning rate, momentum
* 연구시 모델 간의 성능을 좀 더 정직하게 비교할 수 있다.
